File: src/lib/expander.ts
import EventEmitter from "events";
import I2C, { Peripheral } from "./i2c";
import { logger } from "../logger";
import { strPad } from "../util/string";

// --- Constants ---
const DEFAULT_ADDRESS = 0x20;
const DIRECTION_CONF = 0x00; // 0 = output, 1 = input
const POLARITY_CONF = 0x02;
const PULL_UP_CONF = 0x0C;
const INTERRUPT_ENABLE = 0x04;
const INTERRUPT_MODE = 0x08;
const INTERRUPT_COMPARE = 0x06;
const INTERRUPT = 0x0E;
const INTERRUPT_VALUE = 0x10;
const INPUT_VALUE = 0x12;
const OUTPUT_VALUE = 0x14;

// --- Types ---
export interface Config {
  address: number;
  interval: number
}

export interface InterruptResult {
  changed: boolean
  value: number
  previous: number
}

const defaultConfig: Config = {
  address: DEFAULT_ADDRESS,
  interval: 1000 / 120
}

// --- Main Class ---
export class Expander extends EventEmitter {
  private iface: Peripheral;
  private config: Config;
  private timer: NodeJS.Timeout;
  private previous: number;

  constructor(bus: I2C, config?: Config) {
    super()

    this.config = { ...defaultConfig, ...config || {} }
    this.iface = bus.wrap(this.config.address)
  }

  public async init(inputs = 0xFFFF) {
    if (!this.iface?.isConnected()) {
      throw new Error("Air sensor not found");
    }

    // Setup pins as input or output
    await this.setDirection(inputs);

    // Set polarity to normal (0xffff)
    await this.setPolarity(0xFFFF);

    // Set all pull-up resistors enabled (0xffff)
    await this.setPullUp(0xFFFF);

    this.previous = await this.read()

    // Set interrupts: enabled=inputs, mode=0x0000 (compare previous), value=0x0000 (comparison value)
    await this.getInterrupts(inputs, 0x0000, this.previous);

    this.timer = setInterval(async () => {
      const res = await this.readInterrupt()

      if (res.changed) {
        logger.info(`Changed: ${format(res.previous)} -> ${format(res.value)}`)
      }
    }, this.config.interval)
  }

  close() {
    clearInterval(this.timer)
  }

  // --- Configuration Methods ---

  public async setDirection(pins: number): Promise<void> {
    await this.iface.writeRegisterU16LE(DIRECTION_CONF, pins);
  }

  public async getDirection(): Promise<number> {
    return this.iface.readRegisterU16LE(DIRECTION_CONF);
  }

  public async setPolarity(pins: number): Promise<void> {
    await this.iface.writeRegisterU16LE(POLARITY_CONF, pins);
  }

  public async getPolarity(): Promise<number> {
    return this.iface.readRegisterU16LE(POLARITY_CONF);
  }

  public async setPullUp(pins: number): Promise<void> {
    await this.iface.writeRegisterU16LE(PULL_UP_CONF, pins);
  }

  public async getPullUp(): Promise<number> {
    return this.iface.readRegisterU16LE(PULL_UP_CONF);
  }

  public async getInterrupts(enabled: number, mode: number, value: number): Promise<void> {
    await this.iface.writeRegisterU16LE(INTERRUPT_ENABLE, enabled);
    await this.iface.writeRegisterU16LE(INTERRUPT_MODE, mode);
    await this.iface.writeRegisterU16LE(INTERRUPT_COMPARE, value);
  }

  public async getInterruptConfig(): Promise<{ enabled: number, mode: number, value: number }> {
    const enabled = await this.iface.readRegisterU16LE(INTERRUPT_ENABLE);
    const mode = await this.iface.readRegisterU16LE(INTERRUPT_MODE);
    const value = await this.iface.readRegisterU16LE(INTERRUPT_COMPARE);
    return { enabled, mode, value };
  }

  // --- I/O Methods ---

  public async read(): Promise<number> {
    return this.iface.readRegisterU16LE(INPUT_VALUE);
  }

  public async write(value: number): Promise<void> {
    await this.iface.writeRegisterU16LE(OUTPUT_VALUE, value);
  }

  public async readInterrupt(): Promise<InterruptResult> {
    const intr = await this.iface.readRegisterU16LE(INTERRUPT);
    const previous = this.previous

    if (intr === 0) {
      return { changed: false, value: previous, previous };
    }

    const value = await this.iface.readRegisterU16LE(INTERRUPT_VALUE);
    this.previous = value
    return { changed: true, value, previous };
  }
}

function format(val: number) {
  return strPad(`${(val >> 8).toString(2)}`, 8, '0', false) + ' ' + strPad(`${(val & 0xFF).toString(2)}`, 8, '0', false)
}

File: hardware/expander/expander.go
// Sparkfun ??
// https://www.sparkfun.com/??
// https://cdn.sparkfun.com/??

package expander

import (
	"errors"
	"time"

	"github.com/vincent99/velocipi-go/hardware/i2c"
)

const (
	DEFAULT_ADDRESS = 0x20

	// Configuration 8-bit register
	CONFIG = 0x0A // General config, only 0x00 supported here.  [bank, mirror, sequential, slew, hw address, int pin open drain, int pin polarity, unused]

	// Pin configuration, 16-bits, [A7 ... A0 B7 ... B0]
	DIRECTION_CONF    = 0x00 // 0 = output, 1 = input
	POLARITY_CONF     = 0x02 // 0 = normal, 1 = reverse
	PULL_UP_CONF      = 0x0C // 0 = disabled, 1 = pull-up resistor enabled
	INTERRUPT_ENABLE  = 0x04 // 0 = disabled, 1 = enabled interrupt on change
	INTERRUPT_MODE    = 0x08 // 0 = compare previous value, 1 = compare against INTERRUPT_COMPARE
	INTERRUPT_COMPARE = 0x06 // Comparison value used when INTERRUPT_MODE on pin=1

	// Read-Only Interrupt status, 16-bits
	INTERRUPT       = 0x0E // 0 = no change, 1 = pin changed
	INTERRUPT_VALUE = 0x10 // Value of all pins at time of interrupt, value will not change until interrupt cleared via read of this or INPUT_VALUE

	// Read-Write Pin status, 16-bits
	INPUT_VALUE  = 0x12 // Read: Current value of all pins, Write: Sets OUTPUT_VALUE register
	OUTPUT_VALUE = 0x14 // Read: Current status of outputs, Write: Sets value for configured DIRECTION_CONF=output pins
)

type Expander struct {
	iface *i2c.I2C
}

type Config struct {
	Address uint8
	Device  string
}

func NewExpander() (*Expander, error) {
	return NewExpanderWithOptions(&Config{})
}

func NewExpanderWithOptions(opt *Config) (*Expander, error) {
	address := opt.Address
	if address == 0 {
		address = DEFAULT_ADDRESS
	}

	iface, err := i2c.New(opt.Device, address)

	if err != nil {
		return nil, err
	}

	v := &Expander{
		iface,
	}

	return v, v.Init()
}

func (v *Expander) Init() error {
	if !v.IsConnected() {
		return errors.New("expander not found")
	}

	if err := v.SetDirection(0xffbf); err != nil {
		return errors.New("expander could not set input direction: " + err.Error())
	}

	if err := v.SetPolarity(0xffbf); err != nil {
		return errors.New("expander could not be set polarity: " + err.Error())
	}

	if err := v.SetPullUp(0xffbf); err != nil {
		return errors.New("expander could not be set pull-up resistors: " + err.Error())
	}

	if err := v.SetInterrupts(0xffbf, 0x0000, 0x0000); err != nil {
		return errors.New("expander could set interrupt: " + err.Error())
	}

	return nil
}

func (v *Expander) IsConnected() bool {
	var buf []byte
	_, err := v.iface.WriteBytes(buf)
	return err == nil
}

// --------

func (v *Expander) SetDirection(pins uint16) error {
	return v.iface.WriteRegisterU16LE(DIRECTION_CONF, pins)
}

func (v *Expander) GetDirection() (uint16, error) {
	return v.iface.ReadRegisterU16LE(DIRECTION_CONF)
}

func (v *Expander) SetPolarity(pins uint16) error {
	return v.iface.WriteRegisterU16LE(POLARITY_CONF, pins)
}

func (v *Expander) GetPolarity() (uint16, error) {
	return v.iface.ReadRegisterU16LE(POLARITY_CONF)
}

func (v *Expander) SetPullUp(pins uint16) error {
	return v.iface.WriteRegisterU16LE(PULL_UP_CONF, pins)
}

func (v *Expander) GetPullUp() (uint16, error) {
	return v.iface.ReadRegisterU16LE(PULL_UP_CONF)
}

func (v *Expander) SetInterrupts(enabled uint16, mode uint16, value uint16) error {
	err := v.iface.WriteRegisterU16LE(INTERRUPT_ENABLE, enabled)
	if err != nil {
		return err
	}

	err = v.iface.WriteRegisterU16LE(INTERRUPT_MODE, mode)
	if err != nil {
		return err
	}

	return v.iface.WriteRegisterU16LE(INTERRUPT_COMPARE, value)
}

func (v *Expander) GetInterruptConfig() (enabled uint16, mode uint16, value uint16, err error) {
	enabled, err = v.iface.ReadRegisterU16LE(INTERRUPT_ENABLE)
	if err != nil {
		return enabled, mode, value, err
	}

	mode, err = v.iface.ReadRegisterU16LE(INTERRUPT_MODE)
	if err != nil {
		return enabled, mode, value, err
	}

	value, err = v.iface.ReadRegisterU16LE(INTERRUPT_COMPARE)
	return enabled, mode, value, err
}

func (v *Expander) Read() (uint16, error) {
	return v.iface.ReadRegisterU16LE(INPUT_VALUE)
}

func (v *Expander) Write(value uint16) error {
	return v.iface.WriteRegisterU16LE(OUTPUT_VALUE, value)
}

func (v *Expander) readInterrupt() (bool, uint16, error) {
	intr, err := v.iface.ReadRegisterU16LE(INTERRUPT)
	if err != nil {
		return false, 0, err
	}

	if intr == 0 {
		return false, 0, nil
	}

	val, err := v.iface.ReadRegisterU16LE(INTERRUPT_VALUE)
	if err != nil {
		return false, 0, err
	}

	return true, val, nil
}

func (v *Expander) Watch() (chan uint16, chan bool) {
	quit := make(chan bool)
	event := make(chan uint16)

	go func(event chan uint16) {
		ticker := time.NewTicker(10 * time.Millisecond)
		for {
			select {
			case <-ticker.C:
				changed, data, err := v.readInterrupt()
				if err != nil {
					close(event)
					close(quit)
				}

				if changed {
					event <- data
				}

			case <-quit:
				ticker.Stop()
				close(event)
				close(quit)
				return
			}
		}

	}(event)

	return event, quit
}
